12-offer-encoding.md

```diff
@@ -0,0 +1,769 @@
+      - MAY follow with a space and more text
+  - if it can supply more than one item for a single invoice
+    - if the minimum quantity is more than 1:
+      - MUST set that minimum in `quantity_min`
+    - if the maximum quantity is known:
+      - MUST set that maximum in `quantity_max`
+    - if neither:
+      - MUST set `quantity_min` to 1 to indicate `quantity` is supported.
+    - if both:
+      - MUST set `quantity_min` less than or equal to `quantity_max`.
+    - MUST NOT set `quantity_min` or `quantity_max` less than 1.
+  - if `send_invoice` is present:
+    - if the offer is for a partial or full refund for a previously-paid
+      invoice:
+      - SHOULD set `refunded_payment_hash` to the `payment_hash` of that
+        invoice.
+  - otherwise:
+    - MUST NOT set `refunded_payment_hash`.
+
+A reader of an offer:
+  - if `features` contains unknown _odd_ bits that are non-zero:
+    - MUST ignore the bit.
+  - if `features` contains unknown _even_ bits that are non-zero:
+    - MUST NOT respond to the offer.
+    - SHOULD indicate the unknown bit to the user.
+  - if `node_id` or `description` is not set:
+    - MUST NOT respond to the offer.
+  - if `signature` is present, but is not a valid signature using `node_id` as described in [Signature Calculation](#signature-calculation):
+    - MUST NOT respond to the offer.
+  - SHOULD gain user consent for recurring payments.
+  - SHOULD allow user to view and cancel recurring payments.
+  - if it uses `amount` to provide the user with a cost estimate:
+    - MUST warn user if amount of actual invoice differs significantly
+        from that expectation.
+  - SHOULD not respond to an offer if the current time is after
+    `absolute_expiry`.
+  - FIXME: more!
+
+## Rationale
+
+A signature is optional, because it makes for a longer string (potentially
+limiting QR code use on low-end cameras); if the offer has an error, no
+invoice will be given (or, for `send_invoice` offers, accepted), since
+the `offer_id` already covers all the non-signature fields.
+
+# Invoice Requests
+
+Invoice Requests are a request for an invoice; the human-readable prefix for
+invoices is `lnr`.
+
+## TLV Fields for `invoice_request`
+
+1. `tlv_stream`: `invoice_request`
+2. types:
+    1. type: 3 (`chain`)
+    2. data:
+        * [`chain_hash`:`chain`]
+    1. type: 4 (`offer_id`)
+    2. data:
+        * [`sha256`:`offer_id`]
+    1. type: 8 (`amount`)
+    2. data:
+        * [`tu64`:`msat`]
+    1. type: 12 (`features`)
+    2. data:
+        * [`...*byte`:`features`]
+    1. type: 32 (`quantity`)
+    2. data:
+        * [`tu64`:`quantity`]
+    1. type: 38 (`payer_key`)
+    2. data:
+        * [`point32`:`key`]
+    1. type: 39 (`payer_note`)
+    2. data:
+        * [`...*utf8`:`note`]
+    1. type: 50 (`payer_info`)
+    2. data:
+        * [`...*byte`:`blob`]
+    1. type: 56 (`replace_invoice`)
+    2. data:
+        * [`sha256`:`payment_hash`]
+    1. type: 240 (`payer_signature`)
```

Comment:
Agreed

Reply:
Hmm, the other two signatures use the node_id, this uses the payer_key.  I guess there's no confusion since there's no node_id in this message at all.

(52 is *inside* the invoice, and doesn't sign the invoice itself, so that's separate).

Reply:
Actually, I've got a PR which accepts both. If deprecated-apis is enabled, it sends the obsolete sig, otherwise the new one.